core/src/main/java/org/infinispan/conflict/impl/StateReceiverImpl.java

```diff
@@ -0,0 +1,238 @@
+      this.rpcManager = rpcManager;
+   }
+
+   @Start
+   public void start() {
+      this.cache.addListener(this);
+      this.cacheName = cache.getName();
+      this.transferTimeout = cache.getCacheConfiguration().clustering().stateTransfer().timeout();
+   }
+
+   @Override
+   public synchronized void stop() {
+      if (trace) log.tracef("Stop called on StateReceiverImpl for cache %s", cacheName);
+      for (SegmentRequest request : requestMap.values())
+         request.cancel(null);
+   }
+
+   @DataRehashed
+   @SuppressWarnings("unused")
+   synchronized void onDataRehash(DataRehashedEvent dataRehashedEvent) {
+      if (dataRehashedEvent.isPre()) {
+         log.debugf("Cancelling all segment requests as rehash has started %s", dataRehashedEvent);
+
+         for (SegmentRequest request : requestMap.values())
+            request.cancel(new CacheException("Cancelling replica request as the owners of the requested " +
+               "segment have changed."));
+      }
+   }
+
+   @Override
+   public synchronized CompletableFuture<List<Map<Address, CacheEntry<K, V>>>> getAllReplicasForSegment(int segmentId, LocalizedCacheTopology topology) {
+      return requestMap.computeIfAbsent(segmentId, id -> new SegmentRequest(id, topology)).requestState();
+   }
+
+   @Override
+   public void receiveState(Address sender, int topologyId, Collection<StateChunk> stateChunks) {
+      if (stateChunks.isEmpty()) {
+         if (trace)
+            log.tracef("Ignoring received state for cache %s from %s because stateChunks are empty", cacheName, sender);
+         return;
+      }
+
+      int segmentId = stateChunks.iterator().next().getSegmentId();
+      SegmentRequest request = requestMap.get(segmentId);
+      if (request == null) {
+         if (trace) log.tracef("Ignoring received state for cache %s because the associated request was completed or cancelled %s", cacheName);
+         return;
+      }
+      request.receiveState(sender, topologyId, stateChunks);
+   }
+
+   Map<K, Map<Address, CacheEntry<K, V>>> getKeyReplicaMap(int segmentId) {
+      return requestMap.get(segmentId).keyReplicaMap;
+   }
+
+   Map<Address, InboundTransferTask> getTransferTaskMap(int segmentId) {
+      return requestMap.get(segmentId).transferTaskMap;
+   }
+
+   InboundTransferTask createTransferTask(int segmentId, Address source, CacheTopology topology) {
+      return new InboundTransferTask(Collections.singleton(segmentId), source, topology.getTopologyId(),
+            rpcManager, commandsFactory, transferTimeout, cacheName, false);
+   }
+
+   class SegmentRequest {
+      final int segmentId;
+      final LocalizedCacheTopology topology;
+      final List<Address> replicaHosts;
+      final Map<K, Map<Address, CacheEntry<K, V>>> keyReplicaMap = new HashMap<>();
+      final Map<Address, InboundTransferTask> transferTaskMap = new HashMap<>();
+      volatile CompletableFuture<List<Map<Address, CacheEntry<K, V>>>> future;
```

Comment:
you're accessing that only in synchronized block, it does not have to be `volatile`